### PR TITLE
Thread client

### PR DESCRIPTION
**Pull request by kkuehlz**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -53,3 +58,55 @@ func (c *SortClient) Sort(sortCriteria []SortCriterion, searchCriteria *imap.Sea
...
+}
+
+// SupportThread returns true if the remote server supports the extension.
+func (c *ThreadClient) SupportThread() (bool, error) {
+	for _, capability := range ThreadCapabilities {
+		ok, err := c.c.Support(capability)
+		if err != nil {
+			return false, err
+		} else if ok {
+			return true, err
```
Comment (emersion) [inline client.go]: `return true, nil`

File: client.go [hunk 2]
```diff
@@ -53,3 +58,55 @@ func (c *SortClient) Sort(sortCriteria []SortCriterion, searchCriteria *imap.Sea
...
+		}
+	}
+	return false, nil
+}
+
+func (c *ThreadClient) thread(uid bool, threadAlgorithm string, searchCriteria *imap.SearchCriteria) ([]*Thread, error) {
```
Comment (emersion) [inline client.go]: We should probably add a type for threading algorithms, with predefined constants.

File: commands.go [hunk 3]
```diff
@@ -13,6 +13,13 @@ type SortCommand struct {
 	SearchCriteria *imap.SearchCriteria
 }
 
+// ThreadCommand is a THREAD command.
+type ThreadCommand struct {
```
Comment (emersion) [inline commands.go]: Can we move this under the `SortCommand` definitions?

File: responses.go [hunk 4]
```diff
@@ -32,3 +36,52 @@ func (r *SortResponse) Handle(resp imap.Resp) error {
...
+func parseThreadResp(fields []interface{}) ([]*Thread, error) {
+	var parent *Thread
+	var siblings []*Thread
+	for _, f := range fields {
+		switch f := f.(type) {
+		case string:
+			if id, err := imap.ParseNumber(f); err == nil {
+				t := Thread{Id: id, Children: nil}
+				if parent == nil {
+					parent = &t
```
Comment (emersion) [inline responses.go]: Seems like this is done anyway after the `if`

File: responses.go [hunk 5]
```diff
@@ -32,3 +36,52 @@ func (r *SortResponse) Handle(resp imap.Resp) error {
...
+	return nil
+}
+
+func parseThreadResp(fields []interface{}) ([]*Thread, error) {
+	var parent *Thread
+	var siblings []*Thread
+	for _, f := range fields {
+		switch f := f.(type) {
+		case string:
+			if id, err := imap.ParseNumber(f); err == nil {
```
Comment (emersion) [inline responses.go]: We can move the statement out of the `if` and use an early return

```go
id, err := imap.ParseNumber(f)
if err != nil {
    return nil, err
}

t := Thread{Id: id, Children: nil}
…
```

File: responses.go [hunk 6]
```diff
@@ -32,3 +36,52 @@ func (r *SortResponse) Handle(resp imap.Resp) error {
...
+					siblings = append(siblings, &t)
+				} else {
+					parent.Children = append(t.Children, &t)
+				}
+				parent = &t
+			} else {
+				return nil, err
+			}
+		case []interface{}:
+			if t, err := parseThreadResp(f); err == nil {
```
Comment (emersion) [inline responses.go]: Same as above, plus missing error handling

File: sortthread.go [hunk 7]
```diff
@@ -34,3 +41,26 @@ func formatSortCriteria(criteria []SortCriterion) interface{} {
...
+}
+
+func (t *Thread) to_str() string {
+	s := strconv.FormatUint(uint64(t.Id), 10)
+	if len(t.Children) > 1 {
+		for _, child := range t.Children {
+			s += fmt.Sprintf(" (%s)", child.to_str())
+		}
+	} else {
+		for _, child := range t.Children {
```
Comment (emersion) [inline sortthread.go]: No need for a `for` loop

File: sortthread.go [hunk 8]
```diff
@@ -34,3 +41,26 @@ func formatSortCriteria(criteria []SortCriterion) interface{} {
...
+type Thread struct {
+	Id       uint32
+	Children []*Thread
+}
+
+func (t *Thread) String() string {
+	return fmt.Sprintf("(%s)", t.to_str())
+}
+
+func (t *Thread) to_str() string {
```
Comment (emersion) [inline sortthread.go]: No snake_case in Go

**Review by emersion — CHANGES_REQUESTED** [review]
Thanks for your PR! Here are a few comments

**Comment by kkuehlz** [comment]
cc @emersion: Addressed the review comments. Please take another look.

File: sortthread.go (hunk 7 above)
Comment (kkuehlz) [reply sortthread.go]: If I don't use the for loop and print `t.Children` it will print out raw interfaces.

Comment (emersion) [reply sortthread.go]: You can check for `len(t.Children) == 1`, and then just use `t.Children[0]`

File: responses.go [hunk 9]
```diff
@@ -32,3 +36,53 @@ func (r *SortResponse) Handle(resp imap.Resp) error {
...
+	var parent *Thread
+	var siblings []*Thread
+	for _, f := range fields {
+		switch f := f.(type) {
+		case string:
+			id, err := imap.ParseNumber(f)
+			if err != nil {
+				return nil, err
+			}
+			t := Thread{Id: id, Children: nil}
```
Comment (emersion) [inline responses.go]: Nit: no need to specify `Children: nil`

**Comment by kkuehlz** [comment]
cc @emersion is this good to go?

File: responses.go [hunk 10]
```diff
@@ -32,3 +36,53 @@ func (r *SortResponse) Handle(resp imap.Resp) error {
...
+		return err
+	} else {
+		r.Threads = threads
+	}
+
+	return nil
+}
+
+func parseThreadResp(fields []interface{}) ([]*Thread, error) {
+	var parent *Thread
```
Comment (emersion) [inline responses.go:55]: I don't think this is going to cut it. We'd need to remember the whole stack of parents to be able to parse correctly

```
(1 2) (3 4)
   ^ parent is set to 2 here
       ^ parent is still 2 here, should be nil
```

I think we need a recursive `parseThread(fields []interface{})` function instead.

File: thread_test.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import "testing"
+
+var threadTests = []struct {
+	name     string
+	expected string
+	thread   []interface{}
+}{
+	{
+		name:     "simple",
+		expected: "(1 2 3 4)",
```
Comment (emersion) [inline thread_test.go]: Honestly I'm not a fan of this representation for tests. It's not clear the algorithm is doing the right thing. I'd prefer an explicit Go `Thread` struct instead (and probably `reflect.DeepEqual`).

With that, `Thread.String` could be dropped altogether.

**Review by emersion — CHANGES_REQUESTED** [review]
I believe the parsing code has bugs.

File: responses.go (hunk 10 above)
Comment (kkuehlz) [reply responses.go:55]: This method is recursive (line 72) unless I am missing something? Those would be two separate interfaces and I think parent would be set to nil once it enters the interface with (3, 4)

Comment (emersion) [reply responses.go:55]: Oh, time for a cup of coffee it seems

File: responses.go [hunk 12]
```diff
@@ -32,3 +36,53 @@ func (r *SortResponse) Handle(resp imap.Resp) error {
...
+			}
+			t := Thread{Id: id}
+			if parent == nil {
+				siblings = append(siblings, &t)
+			} else {
+				parent.Children = append(t.Children, &t)
+			}
+			parent = &t
+		case []interface{}:
+			t, err := parseThreadResp(f)
```
Comment (emersion) [inline responses.go:72]: Lists are parsed differently when they are children and when they are toplevel. For instance:

```
1 2 3
```

This is three threads.

```
(1 2 3)
```

This is a single thread. 1 has a single child: 2. 2 has a single child: 3.

See the RFC example.

File: thread_test.go [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+	{
+		name:     "noparent",
+		expected: "(3)(5)",
+		thread:   []interface{}{[]interface{}{"3"}, []interface{}{"5"}},
+	},
+	{
+		name:     "nested",
+		expected: "(4 5 (6) (7 8))",
+		thread:   []interface{}{"4", "5", []interface{}{"6"}, []interface{}{"7", "8"}},
+	},
```
Comment (emersion) [inline thread_test.go:138]: Would be nice to include the example from the RFC: `(2)(3 6 (4 23)(44 7 96))`

File: responses.go (hunk 12 above)
Comment (kkuehlz) [reply responses.go:72]: The first example you gave would look like this (1)(2)(3). 

I also have tests for both of these cases. Though I agree the tests need to be reworked as it is not easy to grok what is going on.

**Comment by jeffa5** [comment]
This seems close to ready, is it likely you'd have time to finish it up soon @keur?

**Comment by kkuehlz** [comment]
@jeffa5 Done. 

cc @emersion I made the suggested changes

**Review by emersion — APPROVED** [review]
Thanks for re-working the tests! LGTM, nice work

**Pull request merged**